### PR TITLE
New: CollapseTable

### PR DESCRIPTION
**Pull request by niedzielski**

### CODE REVIEW

File: test/CollapseTable.js [hunk 1]
```diff
@@ -48,4 +48,104 @@ describe('CollapseTable', () => {
...
+                            style='visibility: hidden'
+                          </span>
+                        </div>
+                        <table style='display: block'></table>
+                        <div id=footer style='display: block'></div>
+                      </div>`
+        this.doc = domino.createDocument(html)
+      })
+
+      it('where the footer is observed, the callback is invoked when the footer is clicked',
```
Comment (niedzielski) [inline test/CollapseTable.js]: I tried a more descriptive style for each test case. Let me know your thoughts on this and I'll make the tests consistent one way or another.

Comment (montehurd) [inline test/CollapseTable.js]: I like the descriptive style! 👍

**Comment by niedzielski** [comment]
@montehurd, ready for review with:

https://gerrit.wikimedia.org/r/#/c/349371/
https://gerrit.wikimedia.org/r/#/c/349373/

File: src/CollapseTable.css [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  line-height: 120%;
+  background-color: #f8f8f8;
+  background-repeat: no-repeat;
+  background-position: 95% 50%;
+  background-size: 16px 16px;
+}
+
+/* The outermost table container header's subcaption text. todo: rename. */
+.app_span_collapse_text {
+  color: #808080;
```
Comment (niedzielski) [inline src/CollapseTable.css]: These colors differ from iOS. I'm not sure any of these styles should be here in the long term (maybe just empty classes or classes with visibility changes).

**Comment by montehurd** [comment]
Made a quick pass - will poke around more tomorrow! Exciting stuff!!!

File: src/transform/CollapseTable.css [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+.content table.infobox {
+  font-size: 100%;
+  border-style: none !important;
+}
+
+table.infobox caption {
+  text-align: center;
+  font-weight: bold;
+}
+
```
Comment (montehurd) [inline src/transform/CollapseTable.css]: Do we need the `figcaption` class I see [here](https://gerrit.wikimedia.org/r/#/c/349373/1/app/src/main/assets/applib.css)?

Comment (niedzielski) [inline src/transform/CollapseTable.css]: I think this is now here: https://github.com/wikimedia/wikimedia-page-library/blob/master/src/transform/WidenImage.css#L41

Comment (montehurd) [inline src/transform/CollapseTable.css]: Ah that's right!

File: src/transform/CollapseTable.js [hunk 4]
```diff
@@ -48,6 +51,182 @@ const getTableHeader = (element, pageTitle) => {
...
+    table.style.display = 'none'
+    header.classList.remove('app_table_collapse_close') // todo: use app_table_collapsed_collapsed
+    header.classList.remove('app_table_collapse_icon') // todo: use app_table_collapsed_icon
+    header.classList.add('app_table_collapsed_open') // todo: use app_table_collapsed_expanded
+    if (caption) {
+      caption.style.visibility = 'visible'
+    }
+    footer.style.display = 'none'
+        // if they clicked the bottom div, then scroll back up to the top of the table.
+    if (this === footer && toggleCollapseClickCallback) {
```
Comment (montehurd) [inline src/transform/CollapseTable.js]: I think `toggleCollapseClickCallback` should be `footerDivClickCallback` here?

File: src/transform/CollapseTable.js [hunk 5]
```diff
@@ -48,6 +51,182 @@ const getTableHeader = (element, pageTitle) => {
...
+  for (let i = 0; i < tables.length; ++i) {
+    const table = tables[i]
+
+    if (elementUtilities.findClosestAncestor(table, '.app_table_container')
+      || !shouldTableBeCollapsed(table)) {
+      continue
+    }
+
+    // todo: this is actually an array
+    const headerText = getTableHeader(table, pageTitle)
```
Comment (montehurd) [inline src/transform/CollapseTable.js]: Do you have an example of a page where the transform ends up using `pageTitle`?

Comment (montehurd) [inline src/transform/CollapseTable.js]: I don't see that parameter in the iOS transform so just curious.

**Comment by montehurd** [comment]
We seem to be missing `[openCollapsedTableIfItContainsElement](https://github.com/wikimedia/wikipedia-ios/blob/develop/www/js/transforms/collapseTables.js#L159)`.

We use it on iOS if a [link target is inside a closed table](https://github.com/wikimedia/wikipedia-ios/blob/develop/www/js/listeners.js#L54).

I'll look for an example...

File: src/transform/CollapseTable.js (hunk 4 above)
Comment (niedzielski) [inline src/transform/CollapseTable.js]: Yes! Fixed and added a test.

File: src/transform/CollapseTable.js (hunk 5 above)
Comment (niedzielski) [inline src/transform/CollapseTable.js]: iOS uses window.pageTitle directly. This should now be passed as a parameter. I believe a collapsed references section (third example in CollapseTable.css) would be an example use.

**Comment by niedzielski** [comment]
@montehurd all comments addressed.

> openCollapsedTableIfItContainsElement

Thank you! I'm not sure how I missed this. I've added the method with some tests, however...
- Name changes to **expand**CollapsedTableIfItContainsElement() for consistency (I don't care if we use open and close or expand and collapse but I do care about consistency!)
- The implementation differs from the original in that ParentNode.firstElementChild is used rather than Node.firstChild since classList is only available on Elements.

File: src/transform/CollapseTable.js [hunk 6]
```diff
@@ -218,9 +218,27 @@ const collapseTables = (document, content, pageTitle, isMainPage, infoboxTitle,
   }
 }
 
+/**
+ * @param  {?Element} element
+ * @return {void}
```
Comment (montehurd) [inline src/transform/CollapseTable.js]: Wondering how we want to deal with things like a comment reminding us where issues can be seen... i.e. for this one:
> See first reference link - the `[A]` in `enwiki > Airplane` article from ~June 2016.

File: src/transform/CollapseTable.js (hunk 5 above)
Comment (montehurd) [inline src/transform/CollapseTable.js]: Hmm I think I'm missing something... if I override this parameter with `ASDFASFDASDFA` I don't seem to be able to find any collapsed or expanded table which shows the `ASDFASFDASDFA` string... I've tried quite a few articles...

**Comment by montehurd** [comment]
`expandCollapsedTableIfItContainsElement` works great!

**Comment by montehurd** [comment]
Oops sorry hit the wrong button!

**Comment by montehurd** [comment]
Good news! The iOS integration looks pretty good: https://github.com/wikimedia/wikipedia-ios/pull/1403

File: src/transform/CollapseTable.js (hunk 5 above)
Comment (niedzielski) [inline src/transform/CollapseTable.js]: Resolved offline + adding some comments.

**Comment by niedzielski** [comment]
:+1: Examples, todos, and comments added. All comments addressed.

**Comment by montehurd** [comment]
THIS IS GREAT!

**Pull request merged**